server/storage/backend/backend.go

```diff
@@ -455,7 +456,61 @@ func (b *backend) Commits() int64 {
 }
 
 func (b *backend) Defrag() error {
-	return b.defrag()
+	err := b.defrag()
+	if err != nil {
+		if b.lg != nil {
+			b.lg.Error("defrag failed, trying to reopen db",
+				zap.Error(err),
+			)
+		}
+
+		// TODO: verify that these are all the locks that must be acquired
+		// I copied this from the defrag() function, assuming this is everything that needs to be done
+		b.batchTx.LockOutsideApply()
+		defer b.batchTx.Unlock()
+
+		// lock database after lock tx to avoid deadlock.
+		b.mu.Lock()
+		defer b.mu.Unlock()
+
+		// block concurrent read requests while resetting tx
+		b.readTx.Lock()
+		defer b.readTx.Unlock()
+
+		// TODO: Is there a way to figure out if the bbolt db is closed or open?
+		// Haven't found one so far, thus I'm using b.db.Begin() and then checking the error it returns.
```

Comment:
With the current failpoint `defragNoSpace` the bbolt db will never be closed. Because all other filesystem operations (e.g. the rename of the temp to new db) seem to call a `Fatal` log function it should be safe to assume that the db is still open if we try to restore? If so this would simplify the restore operation a lot and we could avoid trying to figure out if the db is still open or not.

Reply:
no longer relevant